### ISSUE TITLE
Subscription stops receiving messages

### ISSUE THREAD

**Issue by sgnatonski** [nats-io/nats.js#364]
Hi, first of all thanks for great work with nats.js.

I'm having troubles determining the issue I'm observing with subscriptions trying to consume messages produced by a publisher (nats.c client publishing messages at rate 20-25 msg/sec). It looks like a problem of slow consumer but I'm unable to figure out how to fix this with nats.js.

I've stripped my code to bare minimum:

```javascript
var messages = [];
function handleData(data) {
    messages.push(data)
    console.log("msg number " + messages.length)
}
function start(){
    nats.subscribe("events", (msg) => {
        handleData(msg.toString());
    });
}
nats.on("connect", start);
nats.on("unsubscribe", start);
```

and looks like the subscription silently stops to process more messages after receiving approx 5000 messages. Adding more logic to handleData function obviously have an effect of even less messages being received before it hangs. Re-subscribing doesn't seem to help. I've already checked:
- if publisher is still publishing, 
- if nats server runs without issues (nothing logged on the server)

I've also made a test with different nats client (nats.net) and haven't observed any problems with freezed subscriptions, so it seems that the problem is isolated to nats.js.

I was thinking about spinning multiple workers on queued subscription to fetch the messages using multiple threads, but that seems to only postpone the problem...

Are there any guidelines how to approach this problem? Is this something that could be addressed in nats.js?

**Comment by aricart** [comment]
@sgnatonski a couple of things you can do first you need to verify if you are getting a slow consumer. One way to do this is to set up the `name` for the client, and then on the server monitoring check if the client was terminated because it was a slow consumer. You'll be able to see your client by name there. For info on how to access the monitoring for the client take a look at https://docs.nats.io/nats-server/configuration/monitoring#monitoring-nats - if you set it up, then you should be able to see your client at http/s://server:port/connz

If your client is a slow consumer, indeed the client will be done - the `close` handler will trigger, and the client is not usable for any subsequent operation, see https://github.com/nats-io/nats.js#events for how to listen for this.

The node clients (javascript clients) are single-threaded. This means that if your client is busy doing something like publishing in a loop in a bench type scenario, it won't be able to pull messages from the server.

The nats.js client should be able to pubsub 1M messages very quickly (second or so in a laptop).

**Comment by aricart** [comment]
You should be able to run this program (fix the location of the require for nats first) like this: 
`node pubsub -s demo.nats.io -c 1000000 xxxxx`

Note the addition of the `yieldTime` option - this will kick in after the publisher has blasted the number of messages you specified, and when processing a subscription. If the subscription loop takes longer than the specified time (1s), it will give some cpu time to other things running in node (as the I/O).

```javascript
/* jslint node: true */
'use strict'

// change '../' to 'nats' when copying
const nats = require('../')
const argv = require('minimist')(process.argv.slice(2))

const url = argv.s || nats.DEFAULT_URI
const creds = argv.creds
const subject = argv._[0]
const count = argv.c || 1
const msg = argv._[1] || ''

if (!subject) {
  console.log('Usage: node-pub  [-s server] [--creds=filepath] <subject> [msg]')
  process.exit()
}

// Connect to NATS server.
const opts = nats.creds(creds) || {}
opts.yieldTime = 1000
const nc = nats.connect(url, opts)
  .on('connect', () => {
    (async () => {
      let i = 0
      nc.subscribe(subject, (msg) => {
        i++
        if (i % 1000 === 0) {
          console.info(`> ${i} messages`)
        }
      })
    })();

    (async () => {
      let i = 0
      for (i = 0; i < count; i++) {
        nc.publish(subject, msg)
        if (i % 1000 === 0) {
          console.info(`< ${i} messages`)
        }
      }
      nc.flush(() => {
        console.log(`Published ${i} messages`)
        process.exit()
      })
    })()
  })

nc.on('error', (e) => {
  console.log('Error [' + nc.currentServer + ']: ' + e)
  process.exit()
})

```

**Comment by aricart** [comment]
If indeed you are getting a slow consumer because the work in your client is taking longer to process, the way to scale horizontally is create multiple node processes that subscribe to the same subject but specify the same `qgroup`. The work of processing the messages will be distributed between all members of the queue.

**Comment by sgnatonski** [comment]
hi @aricart , thank you so much for such detailed response.

I've followed your advices and ideed your pubsub example works as expected, so I've imported yieldTime option to my code, I've also made sure only 1 publisher and 1 subscriber are connected to nats server. Unfortunately I found no improvement...

i was monitoring the connz endpoint during message processing and I found something interesting: while my node script already stopped receiving new messages (stopped at msg no 5191) nats server was still reporting that the subscriber is receiving messages:

```
{
"server_id": "NBVKB7GTOB6JPMT4EX3HIKI6G5SLKRJQFW3DZXAPUMAP4T4K2WGCNO3F",
"now": "2020-07-16T09:42:16.59244183Z",
"num_connections": 2,
"total": 2,
"offset": 0,
"limit": 1024,
"connections": [
{
"cid": 2,
"ip": "",
"port": 58512,
"start": "2020-07-16T09:36:58.466108831Z",
"last_activity": "2020-07-16T09:42:16.579805742Z",
"rtt": "1.793102ms",
"uptime": "5m18s",
"idle": "0s",
"pending_bytes": 0,
"in_msgs": 5631,
"out_msgs": 0,
"in_bytes": 147440,
"out_bytes": 0,
"subscriptions": 1,
"lang": "C",
"version": "2.1.0"
},
{
"cid": 3,
"ip": "",
"port": 34583,
"start": "2020-07-16T09:37:10.716620892Z",
"last_activity": "2020-07-16T09:42:16.579805742Z",
"rtt": "1.096764ms",
"uptime": "5m5s",
"idle": "0s",
"pending_bytes": 0,
"in_msgs": 0,
"out_msgs": 5563,
"in_bytes": 0,
"out_bytes": 145953,
"subscriptions": 1,
"lang": "node",
"version": "1.4.9"
}
]
}
```

As you can see, the out_msgs was at 5563, while my client hangs at 5191. Observing the difference between in_msgs and out_msgs also suggests that subscriber can't catch up with publisher (the difference in my case is constant ~70 msgs, probably that amount piled up before I started the subscriber), but on the other hand it doesn't look like slow consumer problem, the messages doesn't pile up on server.
What concerns me is the freeze in receiving messages while nats server reports the messages flowing to the subscriber without interruption.
Could it be some rare case, when subscriber (nats.js) receives steady stream of messages but can't catch up with most recent ones so it decides to stop notifying of new messages?

**Comment by sgnatonski** [comment]
I've also split your example into 2 separate scripts (1 for pub, 1 for sub):

```
/* jslint node: true */
'use strict'

// change '../' to 'nats' when copying
const nats = require('nats')
const argv = require('minimist')(process.argv.slice(2))

const url = argv.s || nats.DEFAULT_URI
const creds = argv.creds
const subject = argv._[0]
const count = argv.c || 1
const msg = argv._[1] || ''

if (!subject) {
  console.log('Usage: node-pubsub  [-s server] [--creds=filepath] <subject> [msg]')
  process.exit()
}

// Connect to NATS server.
const opts = nats.creds(creds) || {}
opts.yieldTime = 1000
const nc = nats.connect(url, opts)
  .on('connect', () => {
    (async () => {
      let i = 0
      for (i = 0; i < count; i++) {
        nc.publish(subject, msg)
        if (i % 1000 === 0) {
          console.info(`< ${i} messages`)
        }
      }
      nc.flush(() => {
        console.log(`Published ${i} messages`)
        process.exit()
      })
    })()
  })

nc.on('error', (e) => {
  console.log('Error [' + nc.currentServer + ']: ' + e)
  process.exit()
})
```

executed as `node natspub -s demo.nats.io -c 100000000 xxxxx`


```
/* jslint node: true */
'use strict'

// change '../' to 'nats' when copying
const nats = require('nats')
const argv = require('minimist')(process.argv.slice(2))

const url = argv.s || nats.DEFAULT_URI
const creds = argv.creds
const subject = argv._[0]
const count = argv.c || 1
const msg = argv._[1] || ''

if (!subject) {
  console.log('Usage: node-pubsub  [-s server] [--creds=filepath] <subject> [msg]')
  process.exit()
}

// Connect to NATS server.
const opts = nats.creds(creds) || {}
opts.yieldTime = 1000
const nc = nats.connect(url, opts)
  .on('connect', () => {
    (async () => {
      let i = 0
      nc.subscribe(subject, (msg) => {
        i++
        if (i % 1000 === 0) {
          console.info(`> ${i} messages`)
        }
      })
    })();
  })

nc.on('error', (e) => {
  console.log('Error [' + nc.currentServer + ']: ' + e)
  process.exit()
})
```

executed as `node natssub -s demo.nats.io -c 100000000 xxxxx'
and I was able to reproduce the issue - subscriber hangs after receiving approx 13000 messages:

`Published 100000000 messages` from natspub stdout
`> 13000 messages` from natssub stdout

Note that in order to reproduce this i had to start subscriber before starting publisher

**Comment by aricart** [comment]
> Note that in order to reproduce this i had to start subscriber before starting publisher

Yes this is the correct usage.

I took your scripts as they are and saved them as bulk-pub/sub and ran them:

![render1594910239592](https://user-images.githubusercontent.com/1032976/87685277-d0085080-c748-11ea-9827-519736e56763.gif)

![render1594910083519](https://user-images.githubusercontent.com/1032976/87684832-51abae80-c748-11ea-91c6-867823d92784.gif)

I even triggered the publisher several times, as you can see, while the subscriber is there.

On the monitoring side, looking at `http://127.0.0.1/connz?state=all`:

you can see the subscriber got 5,000,000 messages, the client runs publishing empty messages were 1.3 seconds for 1M.

```json
{
  "server_id": "NB57VJ4XE7DYG32NJFECYU3WSALKXPF2FZ7XWTQ55LNLG5SCLAZQQ2QO",
  "now": "2020-07-16T09:52:56.174801-05:00",
  "num_connections": 6,
  "total": 6,
  "offset": 0,
  "limit": 1024,
  "connections": [
    {
      "cid": 1,
      "ip": "127.0.0.1",
      "port": 63217,
      "start": "2020-07-16T09:52:36.527627-05:00",
      "last_activity": "2020-07-16T09:52:52.525941-05:00",
      "rtt": "655µs",
      "uptime": "19s",
      "idle": "3s",
      "pending_bytes": 0,
      "in_msgs": 0,
      "out_msgs": 5000000,
      "in_bytes": 0,
      "out_bytes": 0,
      "subscriptions": 1,
      "lang": "node",
      "version": "1.4.10-0"
    },
    {
      "cid": 2,
      "ip": "127.0.0.1",
      "port": 63219,
      "start": "2020-07-16T09:52:42.374052-05:00",
      "last_activity": "2020-07-16T09:52:43.608614-05:00",
      "stop": "2020-07-16T09:52:43.614676-05:00",
      "reason": "Client Closed",
      "rtt": "1.915141ms",
      "uptime": "1s",
      "idle": "0s",
      "pending_bytes": 0,
      "in_msgs": 1000000,
      "out_msgs": 0,
      "in_bytes": 0,
      "out_bytes": 0,
      "subscriptions": 0,
      "lang": "node",
      "version": "1.4.10-0"
    },
    {
      "cid": 3,
      "ip": "127.0.0.1",
      "port": 63220,
      "start": "2020-07-16T09:52:45.157075-05:00",
      "last_activity": "2020-07-16T09:52:46.309137-05:00",
      "stop": "2020-07-16T09:52:46.315492-05:00",
      "reason": "Client Closed",
      "rtt": "1.892292ms",
      "uptime": "1s",
      "idle": "0s",
      "pending_bytes": 0,
      "in_msgs": 1000000,
      "out_msgs": 0,
      "in_bytes": 0,
      "out_bytes": 0,
      "subscriptions": 0,
      "lang": "node",
      "version": "1.4.10-0"
    },
    {
      "cid": 4,
      "ip": "127.0.0.1",
      "port": 63221,
      "start": "2020-07-16T09:52:47.344883-05:00",
      "last_activity": "2020-07-16T09:52:48.505028-05:00",
      "stop": "2020-07-16T09:52:48.511269-05:00",
      "reason": "Client Closed",
      "rtt": "1.978878ms",
      "uptime": "1s",
      "idle": "0s",
      "pending_bytes": 0,
      "in_msgs": 1000000,
      "out_msgs": 0,
      "in_bytes": 0,
      "out_bytes": 0,
      "subscriptions": 0,
      "lang": "node",
      "version": "1.4.10-0"
    },
    {
      "cid": 5,
      "ip": "127.0.0.1",
      "port": 63222,
      "start": "2020-07-16T09:52:49.359344-05:00",
      "last_activity": "2020-07-16T09:52:50.500425-05:00",
      "stop": "2020-07-16T09:52:50.506815-05:00",
      "reason": "Client Closed",
      "rtt": "1.946263ms",
      "uptime": "1s",
      "idle": "0s",
      "pending_bytes": 0,
      "in_msgs": 1000000,
      "out_msgs": 0,
      "in_bytes": 0,
      "out_bytes": 0,
      "subscriptions": 0,
      "lang": "node",
      "version": "1.4.10-0"
    },
    {
      "cid": 6,
      "ip": "127.0.0.1",
      "port": 63224,
      "start": "2020-07-16T09:52:51.407721-05:00",
      "last_activity": "2020-07-16T09:52:52.525941-05:00",
      "stop": "2020-07-16T09:52:52.538788-05:00",
      "reason": "Client Closed",
      "rtt": "1.90641ms",
      "uptime": "1s",
      "idle": "0s",
      "pending_bytes": 0,
      "in_msgs": 1000000,
      "out_msgs": 0,
      "in_bytes": 0,
      "out_bytes": 0,
      "subscriptions": 0,
      "lang": "node",
      "version": "1.4.10-0"
    }
  ]
}
```

**Comment by aricart** [comment]
@sgnatonski how big are your payloads?

**Comment by sgnatonski** [comment]
Payload is usually 20-23 bytes, rarely it exceeds 100 bytes.

I see that you are using different nats.js version (1.4.10-0), while in my case its 1.4.9?

**Comment by sgnatonski** [comment]
One more thing - during my regular execution I'm also doing request-reply from subscriber process to publisher process - in this case the payload is much bigger, approx 100KB. Somehow it has a negative impact on the amount of messages received by subscription.
In such case subscription callback stops to being called after few hundred messages

**Comment by sgnatonski** [comment]
Alright, I think I've narrowed it down a bit:

- I've executed my code on Debian machine - runs flawlesly
- if on Debian there are no issues I moved to Win10 machine (on which I observed the issue initialy) and started the code through VS Code with debugger attached - subscription is again stopped at 5191 messages
- I've then ran the code on the same machine (Win10) without debugger - no issue with subscription callback so far

@aricart, looks like issue is caused by simply attaching debugger to node process. While I don't think it can be a problem in production environments I guess it can be problematic during development
